torch/lib/c10d/reducer.cpp

```diff
@@ -1486,6 +1495,13 @@ void Reducer::finalize_backward() {
     local_used_maps_reduced_ = false;
   }
 
+  // Reset bucket counting to ensure we can appropriately launch allreduce for
+  // each bucket in the next backwards.
+  reset_bucket_counting();
```

Comment:
why this needs to be moved out of prepare_backward() and called at multiple places instead?

Reply:
Discussed offline, added a helper function to call this at finalize_backward() and rebuild_buckets() where it is required.